site/app/controllers/course/CourseMaterialsController.php

```diff
@@ -339,15 +339,26 @@ public function ajaxUploadCourseMaterialsFiles() {
                     //cannot check if there are duplicates inside zip file, will overwrite
                     //it is convenient for bulk uploads
                     if ($expand_zip == 'on' && $is_zip_file === true) {
+                        //get the file names inside the zip to write to the JSON file
+                        $zip = zip_open($uploaded_files[1]["tmp_name"][$j]);
+                        while ($zip_entry = zip_read($zip)) {
+                            $zfiles[] =  zip_entry_name($zip_entry);
+                        }
+                        zip_close($zip);
+
                         $zip = new \ZipArchive();
                         $res = $zip->open($uploaded_files[1]["tmp_name"][$j]);
                         if ($res === true) {
                             $zip->extractTo($upload_path);
```

Comment:
Change this to be:
```
                            $entries = [];
                            $disallowed_folders = [".svn", ".git", ".idea", "__macosx"];
                            $disallowed_files = ['.ds_store'];
                            for ($i = 0; $i < $zip->numFiles; $i++) {
                                $entries = $zip->statIndex($i);
                            }
                            $entries = array_filter($entries, function($entry) use ($disallowed_folders, $disallowed_files) {
                                $name = strtolower($entry['name']);
                                foreach ($disallowed_folders as $folder) {
                                    if (Utils::startsWith($folder, $name)) {
                                        return false;
                                    }
                                }
                                if (substr($name, -1) !== '/') {
                                    foreach ($disallowed_files as $file) {
                                        if (basename($name) === $file) {
                                            return false;
                                        }
                                    }
                                }
                                return true;
                            });
                            $zfiles = array_filter($entries, function($entry)) {
                                return substr($entry['name'], -1) !== '/';
                            }
                            $zip->extractTo($upload_path, $entries);
```

This should fix #4475 and not add an entry in `$json` for each folder.